core/src/main/java/hudson/PluginManager.java

```diff
@@ -1949,30 +1950,58 @@ public HttpResponse doUploadPlugin(StaplerRequest req) throws IOException, Servl
     @Restricted(NoExternalUse.class)
     @RequirePOST public FormValidation doCheckUpdateSiteUrl(StaplerRequest request, @QueryParameter String value) {
         Jenkins.get().checkPermission(Jenkins.ADMINISTER);
-        if (StringUtils.isNotBlank(value)) {
-            try {
-                value += ((value.contains("?")) ? "&" : "?") + "version=" + Jenkins.VERSION + "&uctest";
-
-                URL url = new URL(value);
+        return checkUpdateSiteURL(value);
+    }
 
+    @Restricted(DoNotUse.class) // visible for testing only
+    FormValidation checkUpdateSiteURL(@CheckForNull String value) {
+        value = Util.fixEmptyAndTrim(value);
+        if (StringUtils.isBlank(value)) {
+            return FormValidation.error(Messages.PluginManager_emptyUpdateSiteUrl());
+        }
+        final URI baseUri;
+        try {
+            baseUri = new URI(value);
+        } catch (URISyntaxException ex) {
+            return FormValidation.error(ex, Messages.PluginManager_invalidUrl());
+        }
+        if ("file".equalsIgnoreCase(baseUri.getScheme())) {
+            File f = new File(baseUri);
+            if (f.isFile()) {
+                return FormValidation.ok();
+            }
+            return FormValidation.error(Messages.PluginManager_connectionFailed());
+        }
+        if ("https".equalsIgnoreCase(baseUri.getScheme()) || "http".equalsIgnoreCase(baseUri.getScheme())) {
+            final URI uriWithQuery;
+            try {
+                if (baseUri.getRawQuery() == null) {
+                    uriWithQuery = new URI(value + "?version=" + Jenkins.VERSION + "&uctest");
+                } else {
+                    uriWithQuery = new URI(value + "&version=" + Jenkins.VERSION + "&uctest");
+                }
+            } catch (URISyntaxException e) {
+                return FormValidation.error(e, Messages.PluginManager_invalidUrl());
+            }
+            try {
                 // Connect to the URL
-                HttpURLConnection conn = (HttpURLConnection) ProxyConfiguration.open(url);
+                HttpURLConnection conn = (HttpURLConnection) ProxyConfiguration.open(uriWithQuery.toURL());
                 conn.setRequestMethod("HEAD");
                 conn.setConnectTimeout(5000);
                 if (100 <= conn.getResponseCode() && conn.getResponseCode() <= 399) {
                     return FormValidation.ok();
                 } else {
                     LOGGER.log(Level.FINE, "Obtained a non OK ({0}) response from the update center",
-                            new Object[]{conn.getResponseCode(), url});
+                            new Object[]{conn.getResponseCode(), uriWithQuery});
```

Comment:
This is inconsistent with the original commit, which logged `baseUri` here:

```suggestion
                            new Object[]{conn.getResponseCode(), baseUri});
```